cpp/src/arrow/compute/exec/spilling_util.cc

```diff
@@ -0,0 +1,483 @@
+        RETURN_NOT_OK(CollectArrayInfo(batch_info, total_size, child.get()));
+
+    // Cleanup the ArrayData
+    array->type.reset();
+    array->length = 0;
+    return Status::OK();
+}
+
+static Status AllocateBuffersForBatch(SpillFile::BatchInfo &batch_info, MemoryPool *pool)
+{
+#ifndef _WIN32
+    size_t iiovec = 0;
+#endif
+    for(ArrayInfo &arr : batch_info.arrays)
+    {
+        for(size_t ibuf = 0; ibuf < 3; ibuf++)
+        {
+            if(arr.sizes[ibuf])
+            {
+                ARROW_ASSIGN_OR_RAISE(
+                    arr.bufs[ibuf],
+                    AllocateBuffer(
+                        arr.sizes[ibuf],
+                        SpillFile::kAlignment, pool));
+#ifndef _WIN32
+                batch_info.ios[iiovec].iov_base = static_cast<void *>(arr.bufs[ibuf]->mutable_data());
+                batch_info.ios[iiovec].iov_len = static_cast<size_t>(arr.sizes[ibuf]);
+                iiovec++;
+#endif
+            }
+        }
+    }
+    return Status::OK();
+}
+
+Status SpillFile::SpillBatch(QueryContext *ctx, ExecBatch batch)
+{
+    if(handle_ == kInvalidHandle)
+    {
+        ARROW_ASSIGN_OR_RAISE(handle_, OpenTemporaryFile());
+    }
+    int64_t total_size = 0;
+    batches_.emplace_back(new BatchInfo);
+    BatchInfo *info = batches_.back();
+    for(int i = 0; i < batch.num_values(); i++)
+    {
+        if (batch[i].is_scalar())
+            return Status::Invalid("Cannot spill a Scalar");
+        RETURN_NOT_OK(CollectArrayInfo(*info, total_size, batch[i].mutable_array()));
+    }
+    info->start = size_;
+    size_ += total_size;
+
+    FileHandle handle = handle_;
+    RETURN_NOT_OK(ctx->ScheduleIOTask(
+                      [this, handle, info, ctx, total_size]()
+                      {
+                          auto mark = ctx->ReportTempFileIO(total_size);
+                          RETURN_NOT_OK(WriteBatch_PlatformSpecific(handle, *info));
+                          if(++batches_written_ == batches_.size() && read_requested_.load())
+                          {
+                              bool expected = false;
+                              if(read_started_.compare_exchange_strong(expected, true))
+                                  return ctx->ScheduleTask([this, ctx]() { return ScheduleReadbackTasks(ctx); });
+                          }
+                          return Status::OK();
+                      }));
+    return Status::OK();
+}
+
+Status SpillFile::ReadBackBatches(
+    QueryContext *ctx,
+    std::function<Status(size_t, ExecBatch)> fn,
+    std::function<Status(size_t)> on_finished)
+{
+    readback_fn_ = std::move(fn);
+    on_readback_finished_ = std::move(on_finished);
+
+    read_requested_.store(true);
+    if(batches_written_ == batches_.size())
+    {
+        bool expected = false;
+        if(read_started_.compare_exchange_strong(expected, true))
+            return ScheduleReadbackTasks(ctx);
+    }
+    return Status::OK();
+}
+    
+Status SpillFile::Cleanup()
```

Comment:
Are we sure Cleanup() will never be called while there are pending IOs for the file handle? This is related to situations where we want to shut down join early because of 'limit' clause or query error.

Reply:
I don't think there's anything like that at the moment, we can add this early shutdown stuff when we get there? We could use the `subscheduler` mechanism that Weston added.